src/v/cloud_topics/batcher/batcher.cc

```diff
@@ -0,0 +1,273 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/batcher/batcher.h"
+
+#include "cloud_topics/batcher/aggregator.h"
+#include "cloud_topics/batcher/serializer.h"
+#include "cloud_topics/errc.h"
+#include "cloud_topics/logger.h"
+#include "cloud_topics/types.h"
+#include "config/configuration.h"
+#include "container/fragmented_vector.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "model/record_batch_reader.h"
+#include "model/timeout_clock.h"
+#include "ssx/sformat.h"
+#include "storage/record_batch_utils.h"
+#include "utils/human.h"
+#include "utils/retry_chain_node.h"
+#include "utils/uuid.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/file-types.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/shard_id.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/timed_out_error.hh>
+#include <seastar/util/defer.hh>
+
+#include <chrono>
+#include <exception>
+#include <iterator>
+
+using namespace std::chrono_literals;
+
+namespace cloud_topics {
+
+template<class Clock>
+batcher<Clock>::batcher(
+  cloud_storage_clients::bucket_name bucket,
+  cloud_io::remote_api<Clock>& remote)
+  : _remote(remote)
+  , _bucket(std::move(bucket))
+  , _upload_timeout(
+      config::shard_local_cfg().cloud_storage_segment_upload_timeout_ms.bind())
+  , _upload_interval(config::shard_local_cfg()
+                       .cloud_storage_upload_loop_initial_backoff_ms
+                       .bind()) // TODO: use different config
+  , _rtc(_as)
+  , _logger(cd_log, _rtc) {}
+
+template<class Clock>
+ss::future<> batcher<Clock>::start() {
+    ssx::spawn_with_gate(_gate, [this] { return bg_controller_loop(); });
+    return ss::now();
+}
+
+template<class Clock>
+ss::future<> batcher<Clock>::stop() {
+    _as.request_abort();
+    co_await _gate.close();
+    co_return;
+}
+
+template<class Clock>
+typename batcher<Clock>::size_limited_write_req_list
+batcher<Clock>::get_write_requests(size_t max_bytes) {
+    vlog(
+      _logger.debug,
+      "get_write_requests called with max_bytes = {}",
+      max_bytes);
+    size_limited_write_req_list result;
+    size_t acc_size = 0;
+    // The elements in the list are in the insertion order.
+    auto it = _pending.begin();
+    for (; it != _pending.end(); it++) {
+        auto sz = it->data_chunk.payload.size_bytes();
+        acc_size += sz;
+        if (acc_size >= max_bytes) {
+            // Include last element
+            it++;
+            break;
+        }
+    }
+    result.ready.splice(result.ready.end(), _pending, _pending.begin(), it);
```

Comment:
nit: I always have to remind myself what the `splice()` function signature is. A comment like "insert the range of elements from `_pending` into the write request list" could help here.